### ISSUE TITLE
conda cannot install from files based custom channels

### ISSUE THREAD

**Issue by jjhelmus** [conda/conda#2781]
Attempting to install a package from a file based custom channel results in a KeyError.  The custom channel was set up using the steps described in the [Custom channel documentation](http://conda.pydata.org/docs/custom-channels.html).

```
[root@42d0710d45b6 /]# conda index /opt/packages/linux-64/
updating index in: /opt/packages/linux-64
[root@42d0710d45b6 /]# ls /opt/packages/linux-64/
numpy-1.11.0-py35_1.tar.bz2  repodata.json  repodata.json.bz2
[root@42d0710d45b6 /]# conda install -c file:///opt/packages --yes numpy
Fetching package metadata .........
Solving package specifications: ..........

Package plan for installation in environment /opt/conda:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    numpy-1.11.0               |           py35_1         6.1 MB  file:///opt/packages

The following NEW packages will be INSTALLED:

    numpy: 1.11.0-py35_1 file:///opt/packages

Pruning fetched packages from the cache ...
Fetching packages ...
numpy-1.11.0-p 100% |#########################################################################################################| Time: 0:00:00 164.40 MB/s
Extracting packages ...
An unexpected error has occurred, please consider sending the                                                                                      |   0%
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda/issues

Include the output of the command 'conda info' in your report.


Traceback (most recent call last):
  File "/opt/conda/bin/conda", line 6, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.5/site-packages/conda/cli/main.py", line 120, in main
    args_func(args, p)
  File "/opt/conda/lib/python3.5/site-packages/conda/cli/main.py", line 127, in args_func
    args.func(args, p)
  File "/opt/conda/lib/python3.5/site-packages/conda/cli/main_install.py", line 69, in execute
    install(args, parser, 'install')
  File "/opt/conda/lib/python3.5/site-packages/conda/cli/install.py", line 407, in install
    execute_actions(actions, index, verbose=not args.quiet)
  File "/opt/conda/lib/python3.5/site-packages/conda/plan.py", line 589, in execute_actions
    inst.execute_instructions(plan, index, verbose)
  File "/opt/conda/lib/python3.5/site-packages/conda/instructions.py", line 137, in execute_instructions
    cmd(state, arg)
  File "/opt/conda/lib/python3.5/site-packages/conda/instructions.py", line 60, in EXTRACT_CMD
    extract(arg)
  File "/opt/conda/lib/python3.5/site-packages/conda/install.py", line 833, in extract
    rec = package_cache()[dist]
KeyError: 'file:///opt/packages::numpy-1.11.0-py35_1'
```

conda info if it helps

```
[root@42d0710d45b6 /]# conda info
Current conda install:

             platform : linux-64
        conda version : 4.1.2
  conda-build version : 1.20.3
       python version : 3.5.1.final.0
     requests version : 2.10.0
     root environment : /opt/conda  (writable)
  default environment : /opt/conda
     envs directories : /opt/conda/envs
        package cache : /opt/conda/pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/linux-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/linux-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : /root/.condarc
    is foreign system : False
```

This appears to have worked in conda 3.19.3. I'll bisect a bit more to see when this first stopped working.

**Comment by jjhelmus** [comment]
My own testing finds that installing packages from a file backed channels works in conda 4.1.0 but the above error appears in version 4.1.1 and 4.1.2.

**Comment by jarondl** [comment]
I have this problem too, and I believe I have found the culprit in a7e113af27e8d1d23b14413c994b51024f3adef5

**Comment by jjhelmus** [comment]
@jarondl I agree, removing the lines added to the `url_channel` function in a7e113a does allow installation from custom file:// channels.  However, I expect reverting this commit with break conda's ability to install references to an explicit file.

**Comment by mcg1969** [comment]
I'm working on a fix now, guys. I do apologize for the regression.

**Comment by kalefranz** [comment]
fixed in #2812
